text/0000-offset_of.md

```diff
@@ -0,0 +1,576 @@
+design work, and is left to the future.
+
+A drawback is that it is unclear how to support these types in the "Nested Field
+Access" proposed above, so in the future should we decide to support one of
+these, a decision may need to be made about the other.
+
+## `memoffset::span_of!` Functionality
+
+The `memoffset` crate has support for a [`span_of!`][spanof] macro (used like
+`memoffset::span_of!(SomeType, some_field)`), which expands to a `Range<usize>`
+indicating which bytes of `SomeType` are from the field `some_field`.
+
+The use case for this is more limited than that of `offset_of!`, so it was
+omitted from this proposal. That said, should this prove sufficiently useful, it
+would be simple to add a similar macro to `core::mem` in the future.
+
+[spanof]: https://docs.rs/memoffset/0.6.5/memoffset/macro.span_of.html
+
+## Support for types with unsized fields
+
+### ... via `offset_of_val!`
+
+Currently, we don't support use with unsized types. That is, `(A, B, ... [T])`
+and/or `(A, B, ..., dyn Foo)`, or their equivalent in structs.
+
+The reason for this is that the offset of the unsized field is not always known,
+such as in the case of the last field in `(Foo, dyn SomeTrait)`, where the
+offset depends on what the concrete type is. Notably, the compiler must read the
+alignment out of the vtable when you access such a field.
+
+This is equivalent to not being able to determine the the size and/or alignment
+of `?Sized` types, where we solve it by making the user provide the instance
+they're interested in, as in `core::mem::{size_of_val, align_of_val}`, so we
+could provide an analogous `core::mem::offset_of_val!($val, $Type, $field)` to
+support this case.
+
+It would be reasonable to add this in the future, but is left out for now.
+
+### ... by only forbidding the edge case
+
+The only case where we currently do *not* know the offset of a field statically
+is when the user has requested the offset of the unsized field, and the unsized
+field is a trait object.
+
+It's possible for us to provide the offset of for:
+
+1. The fields before the unsized field, as in `offset_of!((i32, dyn Send), 0)`.
+
+2. The unsized field itself if it is a type which whose offset is known without
+   reading the metadata, such as `[T]`, `str`, and types that end with them, as
+   in `offset_of!((i32, [u16]), 1)`, or `offset_of!((u16, (i64, str)), 2)`.
+
+Allowing these is somewhat inconsistent with `core::mem::align_of`, which could
+provide the alignment in some cases such as slices, but instead you must use
+`core::mem::align_of_val` for all `?Sized` types (admittedly, allowing
+`align_of::<[T]>()` is perhaps not very compelling, as it's always the same as
+`align_of::<T>()`).
+
+Either way, it's trivially backwards compatible for us to eventually start
+allowing these, and for the trailing slice/str case, it seems difficult to pin
+down the cases where it's allowed without risk of complicating potential future
```

Comment:
Support in unsized structs is a big blocker for some of my users, and would personally be a big selling point in putting `offset_of` into the language, as that's the only place you can have it.

See also the discussion in: https://github.com/Gilnaa/memoffset/issues/25

Reply:
I agree that support for unsized structs would be great and should be included in the RFC as well.

Reply:
@est31 Do you have a suggestion for how this should work? I think I document the reasons this is hard. Just support it for slices?

Reply:
> Do you have a suggestion for how this should work?

This would need builtin support I think because there is no way to implement this unsound code free. I'm not sure but probably something like having your own AST node [like it was proposed for `format_args`](https://github.com/rust-lang/compiler-team/issues/541). One could also expose it differently, e.g. via a perma unstable feature that is only made to help with implementing the macro. It can't really be a *function* though as functions don't take single fields.

Even if it's, after merging of the RFC, decided to keep it out of the initial version of the macro, it is much easier to add it later through a separate feature gate if it's already been specified as part of an RFC.